typescript/src/redemption.ts

```diff
@@ -0,0 +1,291 @@
+ * transferring Bitcoins from the wallet's main UTXO to the redeemer addresses
+ * and broadcasting it. The change UTXO resulting from the transaction becomes
+ * the new main UTXO of the wallet.
+ * @dev It is up to the caller to ensure the wallet key and each of the redeemer
+ *      addresses represent a valid pending redemption request in the Bridge.
+ *      If this is not the case, an exception will be thrown.
+ * @param bitcoinClient - The Bitcoin client used to interact with the network
+ * @param bridge - The handle to the Bridge on-chain contract
+ * @param walletPrivateKey = The private kay of the wallet in the WIF format
+ * @param mainUtxo - The main UTXO of the wallet. Must match the main UTXO
+ *        held by the on-chain Bridge contract.
+ * @param redeemerAddresses - The list of redeemer addresses
+ * @param witness - The parameter used to decide about the type of the change
+ *        output. P2WPKH if `true`, P2PKH if `false`
+ * @returns Empty promise.
+ */
+export async function makeRedemptions(
+  bitcoinClient: BitcoinClient,
+  bridge: Bridge,
+  walletPrivateKey: string,
+  mainUtxo: UnspentTransactionOutput,
+  redeemerAddresses: string[],
+  witness: boolean
+): Promise<void> {
+  const rawTransaction = await bitcoinClient.getRawTransaction(
+    mainUtxo.transactionHash
+  )
+
+  const mainUtxoWithRaw: UnspentTransactionOutput & RawTransaction = {
+    ...mainUtxo,
+    transactionHex: rawTransaction.transactionHex,
+  }
+
+  const redemptionRequests = await prepareRedemptionRequests(
+    bridge,
+    walletPrivateKey,
+    redeemerAddresses
+  )
+
+  const transaction = await createRedemptionTransaction(
+    walletPrivateKey,
+    mainUtxoWithRaw,
+    redemptionRequests,
+    witness
+  )
+
+  // Note that `broadcast` may fail silently (i.e. no error will be returned,
+  // even if the transaction is rejected by other nodes and does not enter the
+  // mempool, for example due to an UTXO being already spent).
+  await bitcoinClient.broadcast(transaction)
+}
+
+/**
+ * Prepares a list of redemption requests based on the provided redeemer
+ * addresses and the wallet key. Details of the redemption requests are fetched
+ * from the provided Bridge on-chain contract handle.
+ * @dev It is up to the caller of this function to ensure that each of the
+ *      addresses represent a valid pending redemption request in the Bridge
+ *      on-chain contract. An exception will be thrown if any of the addresses
+ *      (along with the wallet public key corresponding to the provided private
+ *      key) does not represent a valid pending redemption.
+ * @param bridge The handle to the Bridge on-chain contract
+ * @param walletPrivateKey The private key of the wallet in the WIF format
+ * @param redeemerAddresses The addresses that will be the recipients of the
+ *        redeemed Bitcoins
+ * @returns The list of redemption requests.
+ */
+async function prepareRedemptionRequests(
```

Comment:
Step 4: Let's refactor this function and work with `redemptionOutputScripts`. I'd also rename that function to `fetchRedemptionRequests` as the requests are not prepared here. Let's make sure all comments and docs refer to the redeemer output scripts instead of addresses (and propagate the changes `index.ts`).

Reply:
Done in https://github.com/keep-network/tbtc-v2/pull/199/commits/4b3be0a5f176ab3e5733f1876ff60fa6d3ca53a2.